File: Array/src/main/java/prefix/Solution121.java
package prefix;

public class Solution121 {
    public int maxProfit(int[] prices) {
        int n = prices.length;

        int[] pre_min = new int[n];
        pre_min[0] = prices[0];

        for (int i = 1; i < n; i++) {
            pre_min[i] = Math.min(pre_min[i - 1], prices[i]);
        }

        int ans = 0;

        for (int i = 0; i < n; i++) {
            ans = Math.max(ans, prices[i] - pre_min[i]);
        }

        return ans;
    }

}

File: Array/src/main/python/solution121.py
from typing import List


class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        n = len(prices)
        pre_min = [0 for _ in range(n)]

        pre_min[0] = prices[0]
        for i in range(1, n):
            pre_min[i] = min(pre_min[i - 1], prices[i])

        ans = 0

        for i in range(1, n):
            ans = max(prices[i] - pre_min[i - 1], ans)

        return ans
